### ISSUE TITLE
Error Transaction Report Filter

### ISSUE THREAD

**Issue by drspoke** [moneymanagerex/moneymanagerex#3748]
MMEX version:
 - [x] 1.5.x
 - [ ] 1.6.x

Note: bug reporters are expected to have verified the bug still exists
either in the last stable version of MMEX or on updated development code
(master branch).

Operating System:
 - [x] Windows
 - [ ] Mac OSX
 - [ ] Linux 

# Description of the bug

I have a problem with the "Transaction Report Filter".
When I make a report only with :
	-	A Specific Account (Bank1)
	-	Status (Reconciled)
	-	Notes (Example*)
The total is incorect.
For example :
If a is tranfer 15$  from the account bank2 to the Bank1, normaly on my Transaction Report Filter, I must see : +15 $
Total 15$.
In this version I see -15$.
Only withdrawals are counted....

I tried the same report on an old version with th e same BDD  and with the 1.3.6 version all was ok ...
Thanks for Your Help

# Reproduction

Is the bug reproducible? 
 - [x] Always 
 - [ ] Randomly 
 - [ ] Happened only once
![Filter Parameters](https://user-images.githubusercontent.com/33481120/128040463-776d0b04-91bc-4f0c-a088-f3a6c6b86078.png)
![New Version Ko](https://user-images.githubusercontent.com/33481120/128040468-b8e71e89-509e-491f-9af6-6d133e6414b0.png)
![Old version1 3 6  OK](https://user-images.githubusercontent.com/33481120/128040473-ba5b29f8-ab6e-4e3b-afbe-f65b4bca79a5.png)

**Comment by whalley** [comment]
Yes, from a quick glance at the code there is also another issue (possibly related) look at here... I think we need to amend some logic now that the support for multiple accounts was added to the transaction filter - https://github.com/moneymanagerex/moneymanagerex/issues/3435 as this checks Transfers in/out against a single account. 

Related also: #3350

**Comment by whalley** [comment]
I guess the question to be asked is how we want to represent TRANSFERS when multiple accounts are specified in the report.

E.g. If we have the following transfers:

A > B $700
B > A $600
A > C $500
C > B $400

and we select the report for Accounts A & B then do we report all transfers to accounts outside the selected accounts as negative and all transfers from accounts outside the selected accounts as positive? What about transfers between accounts in the selected set?

So would this be

A > B +$700   ??
B > A + $700  ??
A > C -$500
C > B +$400

Views?

**Comment by 2BeeOrNot** [comment]
You asked. Here is my view:
Yes. Yes. Show two transactions, one for the transfer to and one for the transfer from.
Savings account:
<img width="448" alt="Savings50" src="https://user-images.githubusercontent.com/63317977/128120778-b737cfbc-a763-4c77-ac4f-2cdba7aa1ad0.png">
Checking account:
<img width="449" alt="Checking50" src="https://user-images.githubusercontent.com/63317977/128120786-3fb6e200-f0d1-4e37-8ff5-66e26d4e3f1d.png">
Savings report. Correct all transactions balance.
<img width="434" alt="ReportSavings50" src="https://user-images.githubusercontent.com/63317977/128120858-ecb76bfc-b46a-49b9-9c0e-c26c7c63b5d2.png">
Checking report. Incorrect all transactions balance. Transfer from Savings should be positive. If it was positive the all transactions balance would be 606.00 instead of -1,094.00. The first transaction should be Account: Checking, Payee: < Savings and Amount: +850.00.
<img width="427" alt="ReportChecking50" src="https://user-images.githubusercontent.com/63317977/128120909-2835438e-fc21-45e2-8d68-aea5da072a07.png">
Checking and Savings report. Incorrect all transactions balance. All transaction balance should be 956.00. There is a negative transfer for the transfer from Savings to Checking but no positive transfer to Checking from Savings. The missing second transfer transaction should be Account: Checking, Payee: < Savings and Amount +850.00. There are only 7 transactions in the report but 8 in the account views - 2 in checking and 6 in Checking.
<img width="425" alt="ReportAll50" src="https://user-images.githubusercontent.com/63317977/128121218-850e1f62-cc7b-4654-ad68-94c4ffa5ce85.png">

**Comment by drspoke** [comment]
Like that the report only takes into account the source account of the transfer and not the recipient account.
which means that all the transfers put hs the report

**Comment by whalley** [comment]
@2BeeOrNot  - Yes, I think that the original logic had the modelling of two transactions in the report for transfers but this is now slightly broken by the multiple account logic. So I think we should report (based on my example above)

```
A > B     +$700  
B > A     -$700
B > A     +$600  
A > B     -$600
A > C     -$500
C > B     +$400
```

i.e. if the from AND to accounts in the transaction are both in the account set being reported then include both transactions otherwise just include the one transaction (fixing the previous noted error on getting the negative/positive right for transfers in).

@drspoke - Having just one transaction does not allow us to see the overall value of transactions  - which was why we introduced the grand totals split.

**Comment by 2BeeOrNot** [comment]
No updated macOS version yet so I turned on the old Windows 7 PC and tried the latest build. It looks good except for one issue. This is the report of the checking account. The amount is now positive like it should be and the total is correct. But shouldn't the Account be 'Checking' and the Payee '< Savings'?
![Transfer-Checking50](https://user-images.githubusercontent.com/63317977/128381116-5008ea65-0f31-4014-96c2-6bf7766b05ec.png)

Like it is for a report of Savings and Checking.
![Transfer-All50](https://user-images.githubusercontent.com/63317977/128381499-ca1fc250-58e3-4f29-ae59-99c8eaaa4295.png)

**Comment by 2BeeOrNot** [comment]
Also related is the All Transactions view. Only one of the transfers is listed. The positive transfer to Checking is not listed.
![AllTrxAllAccounts70](https://user-images.githubusercontent.com/63317977/128384651-9348a2cb-0c25-4db1-ae8f-cef1e0d30bb4.png)

**Comment by whalley** [comment]
> No updated macOS version yet so I turned on the old Windows 7 PC and tried the latest build. It looks good except for one issue. This is the report of the checking account. The amount is now positive like it should be and the total is correct. But shouldn't the Account be 'Checking' and the Payee '< Savings'?
> ![Transfer-Checking50](https://user-images.githubusercontent.com/63317977/128381116-5008ea65-0f31-4014-96c2-6bf7766b05ec.png)
> 
> Like it is for a report of Savings and Checking.
> ![Transfer-All50](https://user-images.githubusercontent.com/63317977/128381499-ca1fc250-58e3-4f29-ae59-99c8eaaa4295.png)

I think the representation is the best that can be done. Internally we hold a single transaction for a transfer with ACCOUNT and TOACCOUNT for transfers.

When we filter we check if the ACCOUNT or TOACCOUNT are in the set of selected accounts, if only one of them are then we report as in your first example. If both accounts are in the set then we report as in your second example duplicating the transaction but reversing the roles of the accounts.

We treat the reporting the same irrespective of whether a single or multiple accounts are selected. 

Interestingly you will note that the that the transaction value for Savings > Check is reversed in the two examples which whilst will end up with the net result is inconsistent so I will look into that.

> Also related is the All Transactions view. Only one of the transfers is listed. The positive transfer to Checking is not listed.

Yes, we only list the actual transaction here and don't need to consider the reverse transactions. There is no concept of balance in this as there is in the report so just listing the actual single transaction is fine.

**Comment by whalley** [comment]
> No updated macOS version yet

@2BeeOrNot Hint taken, new MacOS build available: https://mmex.ipx.co.uk/

**Comment by howff** [comment]
Does this discussion also help with #3350 ?
(Category reports should consider transfers)
